### PR TITLE
Install instructions for CoreDNS in kubeadm

### PR DESCRIPTION
**Pull request by rajansandeep**

### CODE REVIEW

**Comment by rajansandeep** [comment]
@johnbelamaric @chrisohaver @miekg @fturib PTAL.

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+Kubernetes 1.9 has recently been launched and it ships with CoreDNS being a part of it. 
+We can now install CoreDNS as the default service discovery via Kubeadm, which is the toolkit to install Kubernetes easily in a single step.
+
+Currently, [CoreDNS is Alpha in Kubernetes 1.9](https://github.com/kubernetes/features/issues/427). We have a roadmap which will make CoreDNS Beta in version 1.10 and eventually be the default DNS, replacing kube-dns.
+
+## Installing CoreDNS in fresh Kubernetes cluster
+In order to install CoreDNS instead of kube-dns for a fresh Kubernetes cluster, we need to use the `feature-gates` flag and set it to `true`. 
+Use the following command to install CoreDNS as default while installing a fresh Kubernetes cluster.
+~~~ text
+# kubeadm init --feature-gates CoreDNS=true
```
Comment (miekg) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: Should we limit this output to just the relevant coredns bits?

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+  namespace: kube-system
+  resourceVersion: "161"
+  selfLink: /api/v1/namespaces/kube-system/configmaps/coredns
+  uid: 30bf0882-e64e-11e7-baf6-0cc47a8055d6
+~~~
+
+>It is important to note that when switching from kube-dns to CoreDNS, the configurations that come with kube-dns (stubzones, federations...) will no longer exist and will switch to a default configuration in CoreDNS.
+>The translation of the configurations from kube-dns shall be supported from the upcoming version of Kubernetes (v1.10) where CoreDNS will be Beta.
+
+We can check if CoreDNS is functioning normally through a few basic `dig` commands:
```
Comment (miekg) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: Don't you need to be in a pod to have this working?

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+2017/12/21 18:01:35 [INFO] linux/amd64, go1.9.1, a04eeb9c
+10.32.0.1 - [21/Dec/2017:18:06:34 +0000] "A IN kube-dns.kube-system.svc.cluster.local. udp 68 false 4096" NOERROR qr,aa,rd,ra 84 104.117µs
+10.32.0.1 - [21/Dec/2017:18:07:48 +0000] "A IN kubernetes.default.svc.cluster.local. udp 66 false 4096" NOERROR qr,aa,rd,ra 82 116.92µs
+10.32.0.1 - [21/Dec/2017:18:08:53 +0000] "PTR IN 10.0.96.10.in-addr.arpa. udp 53 false 4096" NOERROR qr,aa,rd,ra 105 97.984µs
+10.32.0.1 - [21/Dec/2017:18:09:10 +0000] "PTR IN 1.0.96.10.in-addr.arpa. udp 52 false 4096" NOERROR qr,aa,rd,ra 102 94.992µs
+10.32.0.1 - [21/Dec/2017:18:10:49 +0000] "A IN kubernetes.default.svc.cluster.local. udp 66 false 4096" NOERROR qr,aa,rd,ra 82 99.664µs
+~~~
+
+## Plugins
+CoreDNS in Kubernetes ships with the following `plugins` enabled:
```
Comment (miekg) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: Please lowercase all plugin names and potentially point to the https://coredns.io/plugin/<plugin> page?

File: content/blog/coredns-in-kubernetes-via-kubeadm.md (hunk 2 above)
Comment (chrisohaver) [reply content/blog/coredns-in-kubernetes-via-kubeadm.md]: If you are on a node, I believe you can route directly to the cluster IPs.

However, wrapping the queries in a pod would also verified that the pods are getting correct DNS policy, and communicating with coredns.

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+
+Below is the `log` for some of the basic queries.
+~~~ text
+# kubectl -n kube-system logs coredns-546545bc84-p4x7k
+.:53
+CoreDNS-1.0.1
+linux/amd64, go1.9.1, a04eeb9c
+2017/12/21 18:01:35 [INFO] CoreDNS-1.0.1
+2017/12/21 18:01:35 [INFO] linux/amd64, go1.9.1, a04eeb9c
+10.32.0.1 - [21/Dec/2017:18:06:34 +0000] "A IN kube-dns.kube-system.svc.cluster.local. udp 68 false 4096" NOERROR qr,aa,rd,ra 84 104.117µs
```
Comment (chrisohaver) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: In the latest default deployment config, we'll no longer log every query received.  This section will confuse those who have the newer config.

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+
+~~~
+
+To verify CoreDNS is running, we can check the pod status, deployment and the configmap in the node.
+
+Check `pod` status:
+~~~ text
+# kubectl -n kube-system get pods -o wide
+NAME                               READY     STATUS    RESTARTS   AGE       IP              NODE
+coredns-546545bc84-ttsh4           1/1       Running   0          5h        10.32.0.61      sandeep2
```
Comment (chrisohaver) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: minor detail: latest version of the deployment config has 2 replicas, only one is shown here. This is a minor discrepancy, and probably would not cause confusion.

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+Kubernetes 1.9 has recently been launched and it ships with CoreDNS being a part of it. 
+We can now install CoreDNS as the default service discovery via Kubeadm, which is the toolkit to install Kubernetes easily in a single step.
+
+Currently, [CoreDNS is Alpha in Kubernetes 1.9](https://github.com/kubernetes/features/issues/427). We have a roadmap which will make CoreDNS Beta in version 1.10 and eventually be the default DNS, replacing kube-dns.
+
+## Installing CoreDNS in fresh Kubernetes cluster
+In order to install CoreDNS instead of kube-dns for a fresh Kubernetes cluster, we need to use the `feature-gates` flag and set it to `true`. 
+Use the following command to install CoreDNS as default while installing a fresh Kubernetes cluster.
+~~~ text
+# kubeadm init --feature-gates CoreDNS=true
```
Comment (fturib) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: Can you isolate the command line, and make it bold to show that is the important command.

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+[apiclient] All control plane components are healthy after 31.502217 seconds
+[uploadconfig] Storing the configuration used in ConfigMap "kubeadm-config" in the "kube-system" Namespace
+[markmaster] Will mark node sandeep2 as master by adding a label and a taint
+[markmaster] Master sandeep2 tainted and labelled with key/value: node-role.kubernetes.io/master=""
+[bootstraptoken] Using token: 4cd282.a826a13b3705a4ec
+[bootstraptoken] Configured RBAC rules to allow Node Bootstrap tokens to post CSRs in order for nodes to get long term certificate credentials
+[bootstraptoken] Configured RBAC rules to allow the csrapprover controller automatically approve CSRs from a Node Bootstrap Token
+[bootstraptoken] Configured RBAC rules to allow certificate rotation for all node client certificates in the cluster
+[bootstraptoken] Creating the "cluster-info" ConfigMap in the "kube-public" namespace
+[addons] Applied essential addon: CoreDNS
```
Comment (fturib) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: Maybe we need to explane these 2 lines : is it installed with CoreDNS or with Kubedns.

NOTE: does it worth to short the output with "..." and have these 2 lines in bold ? (the important once).

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+[addons] Applied essential addon: kube-proxy
+
+[upgrade/successful] SUCCESS! Your cluster was upgraded to "v1.10.0-alpha.1". Enjoy!
+
+[upgrade/kubelet] Now that your control plane is upgraded, please proceed with upgrading your kubelets in turn.
+
+~~~
+
+To verify CoreDNS is running, we can check the pod status, deployment and the configmap in the node.
```
Comment (fturib) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: Should be good to talk about the 'kube-dns' service that remain unchanged whatever kube-dns or CoreDNS.

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+NAME      DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
+coredns   1            1          1           1        4h
+~~~
+
+Check the `Configmap`:
+~~~ text
+# kubectl -n kube-system get configmap coredns -oyaml
+apiVersion: v1
+data:
+  Corefile: |
```
Comment (fturib) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: Make a link to the plugin paragraph to explain what is the default configuration, and what is the purpose.

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+metadata:
+  creationTimestamp: 2017-12-21T12:55:15Z
+  name: coredns
+  namespace: kube-system
+  resourceVersion: "161"
+  selfLink: /api/v1/namespaces/kube-system/configmaps/coredns
+  uid: 30bf0882-e64e-11e7-baf6-0cc47a8055d6
+~~~
+
+>It is important to note that when switching from kube-dns to CoreDNS, the configurations that come with kube-dns (stubzones, federations...) will no longer exist and will switch to a default configuration in CoreDNS.
```
Comment (fturib) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: I guess we should say also that CoreDNS is complely compatible with kube-dns. We can point to existing document.

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 11]
```diff
@@ -10,11 +10,33 @@ We can now install CoreDNS as the default service discovery via Kubeadm, which i
 
 Currently, [CoreDNS is Alpha in Kubernetes 1.9](https://github.com/kubernetes/features/issues/427). We have a roadmap which will make CoreDNS Beta in version 1.10 and eventually be the default DNS, replacing kube-dns.
 
+>It is important to note that currently when switching from kube-dns to CoreDNS, the configurations that come with kube-dns (stubzones, federations...) will no longer exist and will switch to a default configuration in CoreDNS.
+>The translation of the configurations from kube-dns shall be supported from the upcoming version of Kubernetes (v1.10) where CoreDNS will be Beta.
+
+
+## [Plugins](https://coredns.io/plugins/)
```
Comment (fturib) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: I would propose to name the chapter "CoreDNS configuration"
add the stanza of the configuration we use for k8s.

explain that configuration of CoreDNS is made of plugin and keep the link to the plugins.

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 12]
```diff
@@ -10,11 +10,33 @@ We can now install CoreDNS as the default service discovery via Kubeadm, which i
...
+CoreDNS in Kubernetes ships with the following `plugins` enabled:
+- *[errors](https://coredns.io/plugins/errors/)*: This enables error logging.
+- *[health](https://coredns.io/plugins/health/)*: health enables a simple health check endpoint.
+- *[kubernetes](https://coredns.io/plugins/kubernetes/)*: The kubernetes plugin enables the reading zone data from a Kubernetes cluster. You can find more details [here](https://coredns.io/plugins/kubernetes/). 
+
+> The `pods insecure` option always return an A record with IP from request (without checking k8s). This option is provided for backward compatibility with kube-dns.
+> Also by default, the Kubernetes plugin has the `Cluster Domain` and the `Service CIDR` defined. The `Pod CIDR` must be added to the config file after CoreDNS deployment.
```
Comment (fturib) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: Hum.. what is the process to update 'Pod CIDR' ? that is not done by the upgrade process ?

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 13]
```diff
@@ -53,34 +75,23 @@ Use the following command to install CoreDNS as default while installing a fresh
...
-Using `kubeadm upgrade plan` and by setting `feature-gates` flag as `true`, it is possible to check the CoreDNS version that will be installed before proceeding to apply the changes.
+Using `kubeadm upgrade plan` and by setting `feature-gates` flag as `CoreDNS=true`, it is possible to check the CoreDNS version that will be installed before proceeding to apply the changes.
 
 Checking the CoreDNS version to upgrade:
-
+~~~ text
+# kubeadm upgrade plan --feature-gates CoreDNS=true
+~~~
 ~~~ text
 # kubeadm upgrade plan --allow-experimental-upgrades --feature-gates CoreDNS=true
```
Comment (fturib) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: I guess you need to be consistent with info 2 lines above or explain why you add the "--allow-experimental-upgrades"

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 14]
```diff
@@ -151,6 +164,7 @@ The upgrade with CoreDNS as the default DNS can then be performed using `kubeadm
 ~~~
 
 To verify CoreDNS is running, we can check the pod status, deployment and the configmap in the node.
```
Comment (fturib) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: I think you should add a title here 'Verifying CoreDNS is providing the service' (or something similar). I guess that will be for whatever process of install-or-upgrade.

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+
+## [Plugins](https://coredns.io/plugins/)
+CoreDNS in Kubernetes ships with the following `plugins` enabled:
+- *[errors](https://coredns.io/plugins/errors/)*: This enables error logging.
+- *[health](https://coredns.io/plugins/health/)*: health enables a simple health check endpoint.
+- *[kubernetes](https://coredns.io/plugins/kubernetes/)*: The kubernetes plugin enables the reading zone data from a Kubernetes cluster. You can find more details [here](https://coredns.io/plugins/kubernetes/). 
+
+> The `pods insecure` option always return an A record with IP from request (without checking k8s). This option is provided for backward compatibility with kube-dns.
+> Also by default, the Kubernetes plugin has the `Cluster Domain` and the `Service CIDR` defined. The `Pod CIDR` must be added to the config file after CoreDNS deployment.
+> `Upstream` option in the Kubernetes plugin defines upstream resolvers to be used resolve external names found (CNAMEs) pointing to external names.
```
Comment (fturib) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: review sentence - a verb is missing I think.

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 16]
```diff
@@ -0,0 +1,268 @@
++++
+title = "Deploying Kubernetes with CoreDNS using kubeadm"
+description = "A guide to installing CoreDNS in Kubernetes via kubeadm"
+tags = ["Kubernetes", "Service", "Discovery", "Kube-DNS", "Kubeadm", "DNS", "Documentation"]
+author = "sandeep"
```
Comment (miekg) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: this needs a date to actually show up on the site

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 17]
```diff
@@ -0,0 +1,268 @@
...
+metadata:
+  creationTimestamp: 2017-12-21T12:55:15Z
+  name: coredns
+  namespace: kube-system
+  resourceVersion: "161"
+  selfLink: /api/v1/namespaces/kube-system/configmaps/coredns
+  uid: 30bf0882-e64e-11e7-baf6-0cc47a8055d6
+~~~
+The Corefile part is the configuration of CoreDNS. 
+This configuration is based on the following [plugins](https://coredns.io/plugins/) of CoreDNS:
```
Comment (miekg) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: this does not render as a list currently, need a line I think.

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 18]
```diff
@@ -0,0 +1,268 @@
...
+- *[errors](https://coredns.io/plugins/errors/)*: Error is logged in stdout.
```
Comment (miekg) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: errors are log to standard output

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 19]
```diff
@@ -0,0 +1,268 @@
...
+- *[health](https://coredns.io/plugins/health/)*: Health check of CoreDNS is reported at http://localhost:8080/health.
```
Comment (miekg) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: health of CoreDNS is reported

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 20]
```diff
@@ -0,0 +1,268 @@
...
+- *[kubernetes](https://coredns.io/plugins/kubernetes/)*: CoreDNS will reply DNS queries based on IP of the services and pods of Kubernetes. You can find more details [here](https://coredns.io/plugins/kubernetes/).
```
Comment (miekg) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: reply to

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 21]
```diff
@@ -0,0 +1,268 @@
...
+  selfLink: /api/v1/namespaces/kube-system/configmaps/coredns
+  uid: 30bf0882-e64e-11e7-baf6-0cc47a8055d6
+~~~
+The Corefile part is the configuration of CoreDNS. 
+This configuration is based on the following [plugins](https://coredns.io/plugins/) of CoreDNS:
+- *[errors](https://coredns.io/plugins/errors/)*: Error is logged in stdout.
+- *[health](https://coredns.io/plugins/health/)*: Health check of CoreDNS is reported at http://localhost:8080/health.
+- *[kubernetes](https://coredns.io/plugins/kubernetes/)*: CoreDNS will reply DNS queries based on IP of the services and pods of Kubernetes. You can find more details [here](https://coredns.io/plugins/kubernetes/). 
+
+> By default, the Kubernetes plugin has its options `Cluster Domain` and `Service CIDR` defined as `cluster.local` and `10.96.0.0/12` respectively. We can modify and choose the desired values through the kubeadm `--service-dns-domain` and `--service-cidr` flags.
```
Comment (miekg) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: what, no?? Or do you mean *if* you use the above config?

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 22]
```diff
@@ -0,0 +1,268 @@
...
+
+> The `pods insecure` option is provided for backward compatibility with kube-dns.
+
+> `Upstream` is used for resolving services that point to external hosts (External Services).
+
+- *[prometheus](https://coredns.io/plugins/prometheus/)*: Metrics of CoreDNS is available at http://localhost:9153/metrics in [Prometheus](https://prometheus.io/) format.
```
Comment (miekg) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: Metrics of CoreDNS are available

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 23]
```diff
@@ -0,0 +1,268 @@
...
+- *[proxy](https://coredns.io/plugins/proxy/)*: Any queries that are not within the cluster domain of Kubernetes will be forwarded to predefined resolvers (/etc/resolv.conf).
+- *[cache](https://coredns.io/plugins/cache/)*: This enables a frontend cache. It will cache all records except zone transfers and metadata records. Increases efficiency.
```
Comment (miekg) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: drop every sentence after "This enabled a frontend cache".

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 24]
```diff
@@ -0,0 +1,268 @@
...
+
+We can modify the default behavior by modifying this configmap. A restart of the CoreDNS pod is required for the changes to take effect. 
+## Installing CoreDNS in fresh Kubernetes cluster
```
Comment (miekg) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: newline above this

File: content/blog/coredns-in-kubernetes-via-kubeadm.md [hunk 25]
```diff
@@ -0,0 +1,268 @@
...
+~~~ text
+# kubectl -n kube-system get deployments
+NAME      DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
+coredns   1            1          1           1        4h
+~~~
+
+We can check if CoreDNS is functioning normally through a few basic `dig` commands:
+
+~~~ text
+# dig @10.32.0.61 kubernetes.default.svc.cluster.local
```
Comment (miekg) [inline content/blog/coredns-in-kubernetes-via-kubeadm.md]: make this shorter with `+noall +answer`, noone cares about all that dig metadata

**Comment by miekg** [comment]
Will put it live tomorrow (UTC+2)

**Comment by rajansandeep** [comment]
Thank you!

**Pull request merged**